File: ice0208/dp-1d/790.py
MOD = 10**9 + 7


class Solution:
    def numTilings(self, n: int) -> int:
        if n == 1:
            return 1
        if n == 2:
            return 2
        if n == 3:
            return 5

        dp = [0 for _ in range(n)]
        dp[0], dp[1], dp[2] = 1, 2, 5

        s = 1
        for i in range(3, n):
            dp[i] = (dp[i - 2] + dp[i - 1] + dp[i - 3] * 2 + s * 2) % MOD
            s += dp[i - 3]

        return dp[-1]

File: moonpoison/dp_1demensional/790.cpp
#include<bits/stdc++.h>
using namespace std;
class Solution {
public:
    int numTilings(int n) 
    {
        long long dp[1001], M=1e9+7;
        dp[1]=1; dp[2]=2; dp[3]=5;
        if(n==1) return 1;
        if(n==2) return 2;
        if(n==3) return 5;
        for(int i=4;i<=n;++i)
            dp[i]=(2*(dp[i-1]%M)+dp[i-3]%M)%M;
        return dp[n];
    }
};